public/app/features/variables/adhoc/AdHocVariableEditor.tsx

```diff
@@ -5,15 +5,13 @@ import { DataSourceRef, SelectableValue } from '@grafana/data';
 
 import { AdHocVariableModel } from '../types';
 import { VariableEditorProps } from '../editor/types';
-import { VariableEditorState } from '../editor/reducer';
-import { AdHocVariableEditorState } from './reducer';
 import { changeVariableDatasource, initAdHocVariableEditor } from './actions';
 import { StoreState } from 'app/types';
 import { VariableSectionHeader } from '../editor/VariableSectionHeader';
 import { VariableSelectField } from '../editor/VariableSelectField';
 
 const mapStateToProps = (state: StoreState) => ({
-  editor: state.templating.editor as VariableEditorState<AdHocVariableEditorState>,
```

Comment:
Could we create a guard like function that takes in state and variable and that returns the correct state type here instead.

Reply:
Yeah, I wanted to take a look at making that easier. I was thinking of something like:

```typescript
function narrowToAdhoc(editorState: VariableEditorState): VariableEditorState<AdhocVariableEditorState> | null {
  return editorState.extended.kind === "adhoc" ? editorState : null; 
}
```

Related: how do we handle the edge case of the AdhocVariableEditor is passed a non-adhoc editor state (somehow??). At the moment it's just continuing on and setting default values, but I'm wondering if it should just "give up" and render null instead? Not sure.

Reply:
If that happens we have problems in other parts of the variable system, so I don't think you need to look into that in this PR.

Reply:
I guess the key part is that if we have problems in our code for the variable system we want to know that before we even run Grafana. That's the whole point of using Typescript. This is just handling one of those edge/error cases.

Reply:
I'm very aware of why we use TypeScript but I think solving a polymorphic architecture with Redux will always be a challenge that comes with some compromises, this is one of them.